File: engine/watcher.py
"""DNS watcher helpers for split-engine."""

from __future__ import annotations

from dataclasses import dataclass

from engine.storage import upsert_domain_observation


@dataclass(slots=True)
class DomainObservation:
    domain: str
    peer: str | None = None
    timestamp: str | None = None


def ingest_dns_event(event: dict) -> DomainObservation | None:
    """Normalize a dnsmasq event into a domain observation and persist it."""
    domain = (event.get("domain") or "").strip().lower().rstrip(".")
    if not domain:
        return None

    observation = DomainObservation(
        domain=domain,
        peer=event.get("peer"),
        timestamp=event.get("timestamp"),
    )
    upsert_domain_observation(observation.domain, peer=observation.peer, seen_at=observation.timestamp)
    return observation

File: internal/watcher/watcher.go
// Package watcher ingests DNS query events and persists them as observations.
package watcher

import (
	"context"
	"strings"
	"time"

	"github.com/belotserkovtsev/split-engine/internal/storage"
)

// Event is a normalized DNS observation.
type Event struct {
	Domain    string
	Peer      string
	Timestamp time.Time
}

// Observation is returned after successful ingest; nil for empty/filtered events.
type Observation struct {
	Domain string
	Peer   string
}

// Ingest normalizes an event and upserts it into storage.
// Returns nil, nil for empty domains.
func Ingest(ctx context.Context, s *storage.Store, e Event) (*Observation, error) {
	domain := strings.TrimRight(strings.ToLower(strings.TrimSpace(e.Domain)), ".")
	if domain == "" {
		return nil, nil
	}

	if err := s.UpsertDomain(ctx, domain, e.Peer, e.Timestamp); err != nil {
		return nil, err
	}
	return &Observation{Domain: domain, Peer: e.Peer}, nil
}
